### PR TITLE
Blog post on the version 1-3 upgrade

### PR DESCRIPTION
**Pull request by NickCuso**
# Description

image

image

image

# Issues

Fixes #
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by NickCuso** [comment]
@nfurfaro had a good question about `grantKeys`...
should we elaborate on the migration steps? particularly around the final 2 steps for locks with keys sold.  I was thinking maybe we could add a screenshot of us calling grantKeys from etherscan with a brief description of what's happening.

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+---
+title: Emergency Upgrade
+subTitle: Version 1-3 is now available
+authorName: Nick Mancuso
+publishDate: January 17, 2020
+description: Explaining what went wrong and why we released a new version so quickly after our last upgrade.
+image: /static/images/blog/unlocking-smart-contracts/code.jpeg
+---
+
+We recently noticed a vulnerability which if abused could have had bad consequences – but we caught it early and fixed it promptly. This post aims to explain the errors we made with our most recent smart-contract update.
```
Comment (julien51) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]:
```suggestion
We recently noticed a vulnerability which if abused could have had bad consequences – but we caught it early and fixed it promptly. This post aims to explain the errors we made with our [most recent smart-contract update](https://unlock-protocol.com/blog/version-1-2/).
```

Comment (julien51) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: Just adding a link ;)

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+There were problems with this release which if targeted by an attacker had the potential to do one of the following:
+ - Steal money made from previous key sales.
+ - Break every lock (and key) using the new version - blocking new sales, preventing key owners from being recognized and preventing the lock owner from accessing any funds previously made.
+
+We fixed the second issue already, so no worries there.  But funds are at risk if your lock is impacted.  For that we need the lock owners themselves to upgrade ASAP.
```
Comment (julien51) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]: it seems to imply that the first issue not fixed?

```suggestion
We fixed the second issue very quickly, so no worries there.  But funds are at risk if your lock is impacted.  For that we need the lock owners themselves to upgrade ASAP.
```

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+This only impacts locks created in the past couple weeks. If you recently created a lock, please go to the Unlock dashboard to check if you are impacted.  The dashboard will flag any impacted lock, asking you to upgrade.
```
Comment (julien51) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]: Let's be more precise on dates (ideally even block numbers!)

Comment (julien51) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: And lets something along the lines of "if the dashboard does not show a warning  like the one below, your lock is safe". 

[I'll send a srceenshot ASAP]

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+We recently changed the locks deployed to use the minimal proxy standard, [EIP-1167](https://eips.ethereum.org/EIPS/eip-1167).  This significantly reduces the cost to create new locks and leads to less blockchain bloat.  When using proxies any data set in the constructor is lost so instead we call a single-use function called `initialize`.
+
+During normal operation a user calls `createLock` on our main Unlock contract.  That function will deploy the lock and call `initialize` in the same transaction.  It’s responsible for recording data such as the lock owner and the key price.
+
+Proxies work by reading the implementation logic from a template.  The template is nothing more than a contract containing the logic / source code for the lock proxies to leverage.
+
+We forgot to `initialize` the template!
+
+If someone figured this out before we did they could have called `initialize` on the template to make themselves the owner and then make an owner-only call to `selfdestruct` the contract.
```
Comment (nfurfaro) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]: I wonder if we should mention why we chose to keep `selfdestruct` in the first place? Minor thing, but it was a conscious decision with the goal of helping minimize bloat on the blockchain.

Comment (nfurfaro) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: openzeppelin even has a warning about this here: https://docs.openzeppelin.com/upgrades/2.6/writing-upgradeable#potentially-unsafe-operations

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+Proxies work by reading the implementation logic from a template.  The template is nothing more than a contract containing the logic / source code for the lock proxies to leverage.
+
+We forgot to `initialize` the template!
+
+If someone figured this out before we did they could have called `initialize` on the template to make themselves the owner and then make an owner-only call to `selfdestruct` the contract.  
+
+Once the template has been destroyed, all proxies attempting to read the implementation logic will fail. This means no new purchases, no reading who owns a key, and no way to withdraw any funds raised from previous key sales.
+
+We initialized the template and renounced ownership so that no one can ever call selfdestruct. Going forward we have added this to our deployment checklist, ensuring the template is initialized before Unlock is upgraded so that there is no window where someone could do damage by claiming ownership first.
```
Comment (nfurfaro) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]: could we add these steps (initialize and renounceOwnership) to the deploy-template script for a more robust fix here? I personally just deployed the template using myCrypto, but would be happy to switch to using a script, especially if we automated these actions. Thenwe could say "Going forward we have added this to our deployment script" which in my mind instills more confidence.

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md (hunk 4 above)
Comment (julien51) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: I think it's worth adding this link to the blog post!

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md (hunk 5 above)
Comment (julien51) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: It's actually wip ;)

Comment (julien51) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: Lets finish a very bold and re-assuring statement saying that this issue has now been fixed and that no one wa able to "abuse" it in any way.

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+## Steal money from key sales
+
+> This issue is due to a bug in our code.  The bug has been fixed so that any new locks created going forward are fine.  But some locks were impacted and we need lock owners to do the upgrade.
```
Comment (julien51) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]: Great!

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+```
+  function initialize() public
+  {
+    // default to 10%
+    refundPenaltyBasisPoints = 1000;
+  }
+```
+
+These two calls are public and have no modifier.  This means anyone can call them at any time.
```
Comment (julien51) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]:
```suggestion
These two calls are public and have no modifier.  This means anyone can call them at any time on affected locks.
```

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md (hunk 3 above)
Comment (julien51) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: image

Comment (NickCuso) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: Will do.  
Issue introduced: https://etherscan.io/tx/0x159252de7b715cb7e773e5eb54435cd19a8563a574cc5ecfd4f6fcff7bc7204f
Issue resolved: https://etherscan.io/tx/0xa66dc730133db01cc39a30ea7f968558e6f4e5aa1305c5d9851e0547810f71c8

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md (hunk 4 above)
Comment (NickCuso) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: Added both, thanks

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md (hunk 5 above)
Comment (NickCuso) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: Added both.

**Comment by NickCuso** [comment]
Feedback addressed and screenshots updated.
Please take another pass, particularly at the ending.

**Comment by NickCuso** [comment]
Pushed the new art and updated screenshots again

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+image: /static/images/blog/emergency-upgrade-1-3/code_vulnerability.jpg
+---
+
+We recently noticed a vulnerability which if abused could have had bad consequences – but we caught it early and fixed it promptly. This post aims to explain the errors we made with our [most recent smart-contract update](https://unlock-protocol.com/blog/version-1-2/).
+
+There were problems with this release which if targeted by an attacker had the potential to do one of the following:
+ - Steal money made from previous key sales.
+ - Break every lock (and key) using the new version - blocking new sales, preventing key owners from being recognized and preventing the lock owner from accessing any funds previously made.
+
+We fixed the second issue very quickly, so no worries there.  But funds are at risk if your lock is impacted.  For that we need the owners of these locks themselves to upgrade ASAP. (there are 14 locks which affected)
```
Comment (julien51) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]:
```suggestion
We fixed the second issue very quickly, so no worries there.  But funds are at risk if your lock is impacted.  For that we need the owners of these locks themselves to upgrade ASAP. (there are 14 locks which are affected)
```

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+This only impacts locks created between Jan 6th until Jan 17th (locks created in blocks #9227829 - #9295958). If you recently created a lock, please go to the Unlock dashboard to check if you are impacted. If the dashboard does not show a warning like the one below, your lock is safe.
```
Comment (julien51) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]:
```suggestion
This only impacts locks created between Jan 6th until Jan 17th (locks created between the block #9227829 and the block #9295958). If you recently created a lock, please go to the Unlock dashboard to check if you are impacted. If the dashboard does not show a warning like the one below, **your lock is safe**.
```

File: unlock-protocol.com/blog/emergency-upgrade-1-3.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+## Break every lock using the new version
+
+> This issue is about our deployment process.  We simply needed to take action before an attacker.  We will avoid this race condition in the future.
+
+We recently changed the locks deployed to use the minimal proxy standard, [EIP-1167](https://eips.ethereum.org/EIPS/eip-1167).  This significantly reduces the cost to create new locks and leads to less blockchain bloat.  When using proxies any data set in the constructor is lost so instead we call a single-use function called `initialize`.
+
+During normal operation a user calls `createLock` on our main Unlock contract.  That function will deploy the lock and call `initialize` in the same transaction.  It’s responsible for recording data such as the lock owner and the key price.
+
+Proxies work by reading the implementation logic from a template.  The template is nothing more than a contract containing the logic / source code for the lock proxies to leverage.
```
Comment (julien51) [inline unlock-protocol.com/blog/emergency-upgrade-1-3.md]:
```suggestion
Proxies work by reading the implementation logic from a template.  The template is nothing more than a contract containing the logic / source code for the lock implementations to leverage.
```

Comment (NickCuso) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: hmm that sounds a little confusing to me.  maybe `... / source code for each individual lock to leverage`

Comment (NickCuso) [reply unlock-protocol.com/blog/emergency-upgrade-1-3.md]: I pushed with the text I suggested.  let me know if this sounds alright.

**Pull request merged**